### PR TITLE
Deprecate `LoggerCollection` in favor of `trainer.loggers`

### LINKED ISSUE #11232: [RFC] Better support using multiple loggers simultaneously by deprecating `LoggerCollection`
## Proposed refactor

Deprecate `LoggerCollection`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/a6a28e08d22f59f2468ff1049c67507df7519e7c/pytorch_lightning/loggers/base.py#L370-L464

## Motivation

The logger collection implements the base logger interface and wraps a sequence of loggers. This way, the trainer always behaves as though a single logger is being used. Such a collection is somewhat handy for writing data, but is much less helpful when reading state from loggers, especially when individual loggers have different state/property values. For instance, It's not clear at all what `LoggerCollection.name` *should* be. The concatenation of all its instance loggers names? The first logger's name? This also came up here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10954

This leads to inconsistencies across components. Namely, Lightning offers no such LoggerCollection-equivalent for Callbacks. The trainer treats them as a flat list, and when hooks need to be called, the Trainer iterates over this list, and calls the callbacks' hooks successively. Why do we not do the same for loggers? 

This leads to leaks in the user-facing API. If users pass a list of loggers to the `Trainer`'s `logger` flag, accessing `trainer.logger` returns a single instance of type `LoggerCollection`, which is different than what users specified, which means users have to learn about this abstraction and likely update their code to handle cases where a single logger is used vs multiple.

## Pitch

1. Introduce a `loggers` property on the Trainer which returns a list of the passed in Loggers objects, so users have an option to not use the LoggerCollection
```py
@property
def loggers(self) -> List[LightningLoggerBase]:
```
2. Deprecate `LoggerCollection` in favor of this `loggers` property. Wherever we call the Logger APIs from within the Trainer, iterate over the `loggers` list and call the hooks instead. 

Example:

Current
https://github.com/PyTorchLightning/pytorch-lightning/blob/a6a28e08d22f59f2468ff1049c67507df7519e7c/pytorch_lightning/trainer/trainer.py#L1444-L1445

New
```py
for logger in self.loggers:
    logger.finalize("success")
```

3. After `LoggerCollection` is removed, we could redefine `logger `on the Trainer to be something like this. Or we could deprecate `logger` entirely in favor of `loggers` to avoid any potential misuse/missed expectations around supporting multiple loggers. 

```py
@property
def logger(self) -> Optional[LightningLoggerBase]:
    loggers = self.loggers
    return loggers[0] if len(loggers) > 0 else None
```

_Originally posted by @ananthsub in https://github.com/PyTorchLightning/pytorch-lightning/issues/11209#issuecomment-999393358_
### Pitch

cc @justusschock @awaelchli @akihironitta @edward-io @borda @ananthsub @tchaton

### PR DESCRIPTION
**Pull request by akashkw**
## What does this PR do?

Fixes #11232

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/properties/test_loggers.py [hunk 1]
```diff
@@ -52,6 +54,10 @@ def test_trainer_loggers_setters():
     logger2 = CustomLogger()
     logger_collection = LoggerCollection([logger1, logger2])
     logger_collection_2 = LoggerCollection([logger2])
+    logger_collection_dep = "Using `trainer.logger` when Trainer is configured to use multiple loggers."
+    " `LoggerCollection` is deprecated in v1.6 in favor of `trainer.loggers` and will be removed in v1.8."
+    " This behavior will change in v1.8 such that `trainer.logger` will return the first"
+    " logger in `trainer.loggers`."
```
Comment (carmocca) [inline tests/trainer/properties/test_loggers.py:60]: Don't need to match the full string, `match=...` takes a regex

```suggestion
    logger_collection_dep = "LoggerCollection` is deprecated in v1.6"
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2599,10 +2599,11 @@ def logger(self) -> Optional[LightningLoggerBase]:
...
         else:
-            rank_zero_warn(
-                "Using trainer.logger when Trainer is configured to use multiple loggers."
-                " This behavior will change in v1.8 when LoggerCollection is removed, and"
-                " trainer.logger will return the first logger in trainer.loggers"
+            rank_zero_deprecation(
+                "Using `trainer.logger` when Trainer is configured to use multiple loggers."
+                " `LoggerCollection` is deprecated in v1.6 in favor of `trainer.loggers` and will be removed in v1.8."
+                " This behavior will change in v1.8 such that `trainer.logger` will return the first"
+                " logger in `trainer.loggers`."
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2606]: Shouldn't this be inside the `LoggerCollection.__init__`?

Comment (akashkw) [reply pytorch_lightning/trainer/trainer.py:2606]: It could be, this is still up for discussion. @daniellepintz and I talked about it earlier and we concluded that since there was already a warning here, it made sense to replace it with a deprecation warning. 

Most people who are using `LoggerCollection` are doing so through `trainer.logger` so they should be made aware of both the deprecation of `LoggerCollection` and the changing behavior of `trainer.logger`.  

What we could do is have one warning inside the `trainer.logger` getter and another warning in `LoggerCollection.__init__`. The `trainer.logger` getter constructs a `LoggerCollection` each time it's called so both warnings would be displayed.

Comment (akashkw) [reply pytorch_lightning/trainer/trainer.py]: For now, having two deprecation warnings makes the most sense to me, so I went ahead and added a deprecation message to `LoggerCollection.__init__`.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2599,10 +2599,10 @@ def logger(self) -> Optional[LightningLoggerBase]:
         if len(self.loggers) == 1:
             return self.loggers[0]
         else:
-            rank_zero_warn(
-                "Using trainer.logger when Trainer is configured to use multiple loggers."
-                " This behavior will change in v1.8 when LoggerCollection is removed, and"
-                " trainer.logger will return the first logger in trainer.loggers"
+            rank_zero_deprecation(
```
Comment (daniellepintz) [inline pytorch_lightning/trainer/trainer.py:2602]: I think this should be `rank_zero_warn` because otherwise the framework will emit dep warnings every time it accesses `trainer.logger` when we have multiple loggers.

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py]: I agree it makes sense to have a warning in both places

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -2600,9 +2600,9 @@ def logger(self) -> Optional[LightningLoggerBase]:
             return self.loggers[0]
         else:
             rank_zero_warn(
-                "Using trainer.logger when Trainer is configured to use multiple loggers."
-                " This behavior will change in v1.8 when LoggerCollection is removed, and"
-                " trainer.logger will return the first logger in trainer.loggers"
+                "Using `trainer.logger` when Trainer is configured to use multiple loggers."
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2603]:
```suggestion
                "You are using `trainer.logger` when the Trainer is configured to use multiple loggers."
```

**Review by ananthsub — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -2600,9 +2600,9 @@ def logger(self) -> Optional[LightningLoggerBase]:
             return self.loggers[0]
         else:
             rank_zero_warn(
-                "Using trainer.logger when Trainer is configured to use multiple loggers."
-                " This behavior will change in v1.8 when LoggerCollection is removed, and"
-                " trainer.logger will return the first logger in trainer.loggers"
+                "You are using `trainer.logger` when the Trainer is configured to use multiple loggers."
+                " This behavior will change in v1.8 such that `trainer.logger` will return the first"
+                " logger in `trainer.loggers`."
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2605]:
```suggestion
                "Using trainer.logger when Trainer is configured to use multiple loggers."
                " This behavior will change in v1.8 when LoggerCollection is removed, and"
                " trainer.logger will return the first logger in trainer.loggers"
```

File: pytorch_lightning/loggers/base.py [hunk 6]
```diff
@@ -221,13 +221,19 @@ def version(self) -> Union[int, str]:
 class LoggerCollection(LightningLoggerBase):
     """The :class:`LoggerCollection` class is used to iterate all logging actions over the given `logger_iterable`.
 
+    .. deprecated:: v1.6
+        `LoggerCollection` is deprecated in v1.6 in favor of `trainer.loggers` and will be removed in v1.8.
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:225]: I would prefer to make the error message more explicit. The short changelog entry is fine but here we need to say what the user should do instead. 

```suggestion
        `LoggerCollection` is deprecated in v1.6 and will be removed in v1.8. Directly pass a list of loggers to the Trainer and access the list via the `trainer.loggers` attribute. 
```

File: pytorch_lightning/loggers/base.py [hunk 7]
```diff
@@ -221,13 +221,19 @@ def version(self) -> Union[int, str]:
...
+
     Args:
         logger_iterable: An iterable collection of loggers
     """
 
     def __init__(self, logger_iterable: Iterable[LightningLoggerBase]):
         super().__init__()
         self._logger_iterable = logger_iterable
+        rank_zero_deprecation(
+            "`LoggerCollection` is deprecated in v1.6 in favor of `trainer.loggers` and will be removed in v1.8."
```
Comment (awaelchli) [inline pytorch_lightning/loggers/base.py:235]: same here :))

File: tests/trainer/properties/test_loggers.py [hunk 8]
```diff
@@ -50,8 +52,10 @@ def test_trainer_loggers_setters():
     """Test the behavior of setters for trainer.logger and trainer.loggers."""
     logger1 = CustomLogger()
     logger2 = CustomLogger()
-    logger_collection = LoggerCollection([logger1, logger2])
-    logger_collection_2 = LoggerCollection([logger2])
+    with pytest.deprecated_call(match="`LoggerCollection` is deprecated in v1.6"):
+        logger_collection = LoggerCollection([logger1, logger2])
+    with pytest.deprecated_call(match="`LoggerCollection` is deprecated in v1.6"):
+        logger_collection_2 = LoggerCollection([logger2])
```
Comment (awaelchli) [inline tests/trainer/properties/test_loggers.py:58]: this seems to be unused, weird.

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/properties/test_loggers.py (hunk 8 above)
Comment (akashkw) [reply tests/trainer/properties/test_loggers.py:58]: Not sure I understand what you mean. Can you clarify?
`logger_collection` and `logger_collection_2` are both used later on in this unit test.

File: pytorch_lightning/loggers/base.py (hunk 6 above)
Comment (daniellepintz) [reply pytorch_lightning/loggers/base.py]: I think this is fine, but the user actually doesn't have to change any behavior, they still pass a list of loggers to the Trainer

Comment (awaelchli) [reply pytorch_lightning/loggers/base.py]: But hold on, this warning is when LoggerCollection gets instantiated by the user. This is what we want to warn about, and passing it to the trainer should trigger a warning as well. However, we need to avoid triggering the warning internally or multiple times. This is currently  a problem, because for the getter in 

`trainer.logger`

we return 

`return LoggerCollection(self.loggers)`

which will once again trigger a warning that the user will find confusing and can't change. Do you agree?
Can we make an internal flag that we flip when we want to suppress the warning internally?

Comment (ananthsub) [reply pytorch_lightning/loggers/base.py]: We should still warn about that scenario because the behavior will change following v1.8 as users will receive the first logger in the list instead of a logger collection. To update their code, they would need to use trainer.loggers

Comment (daniellepintz) [reply pytorch_lightning/loggers/base.py]: @awaelchli after https://github.com/PyTorchLightning/pytorch-lightning/pull/11920 there are virtually no places where we call `trainer.logger` internally - we updated them all to `trainer.loggers`

Comment (awaelchli) [reply pytorch_lightning/loggers/base.py]: @daniellepintz To be very clear:

```py
from pytorch_lightning import Trainer
from pytorch_lightning.loggers import TensorBoardLogger


if __name__ == "__main__":
    trainer = Trainer(
        logger=[TensorBoardLogger(save_dir="."), TensorBoardLogger(save_dir=".")]
    )
    _ = trainer.logger
```

Nothing in the above code is deprecated usage. Yet we see this message:

```
/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/trainer/trainer.py:2621: UserWarning: Using trainer.logger when Trainer is configured to use multiple loggers. This behavior will change in v1.8 when LoggerCollection is removed, and trainer.logger will return the first logger in trainer.loggers
  rank_zero_warn(
/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/loggers/base.py:235: LightningDeprecationWarning: `LoggerCollection` is deprecated in v1.6 and will be removed in v1.8. Directly pass a list of loggers to the Trainer and access the list via the `trainer.loggers` attribute.
  rank_zero_deprecation(
```

There is nothing I as a user can do to avoid the second warning, except change to `trainer.loggers` but that's not deprecated API. Plus, the warning mentions `LoggerCollection` which I have not used anywhere in my code. 

It's triggered internally because we instantiate `LoggerCollection` and we need to suppress that warning, for example with https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings or the toggle as I showed above. No matter how unlikely it is that a user runs into this issue, we can't do it because it's not correct.

Comment (akashkw) [reply pytorch_lightning/loggers/base.py]: We were actually having a similar discussion earlier. https://github.com/PyTorchLightning/pytorch-lightning/pull/12147#r816308356. Having the user see both the warning and the deprecation was an intentional choice (originally both messages were combined into one). I agree though that it would be confusing for the user to see the deprecation warning from instantiating `LoggerCollection` so this is a good solution
